lib/game/distributed.go

```diff
@@ -0,0 +1,221 @@
+package game
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"strings"
+)
+
+const (
+	IDLength = 32
+)
+
+var NullID = strings.Repeat("0", IDLength)
+
+type DistributedUniverse struct {
+	*Universe
+
+	ID             string
+	TopID          string
+	TopNeighbor    *DistributedUniverse `json:"-"`
+	BottomID       string
+	BottomNeighbor *DistributedUniverse `json:"-"`
+	LeftID         string
+	LeftNeighbor   *DistributedUniverse `json:"-"`
+	RightID        string
+	RightNeighbor  *DistributedUniverse `json:"-"`
+
+	// set the GetNeighbor function with implementation for
+	// the distributed environment.
+	GetNeighbor func(id string) *DistributedUniverse `json:"-"`
+}
+
+func NewDistributedUniverse(id string, height, width uint32) *DistributedUniverse {
+	d := &DistributedUniverse{
+		ID:       id,
+		TopID:    NullID,
+		BottomID: NullID,
+		LeftID:   NullID,
+		RightID:  NullID,
+		Universe: NewUniverse(height, width),
+	}
+
+	d.Rules = d.rules
+	d.GetNeighbor = d.getEmptyUniverse
```

Comment:
What if we require this function to be passed? Does it make sense to have a distributed universe without being able to get neighbors?

Reply:
The default makes the universe act like it has no neighbors. That way you do not have to hookup that function just to load or save a universe.